### PR TITLE
Modules Design update

### PR DESCRIPTION
**Pull request by eriqua**
# Change

Adding bicep modules section and deployment names subsection
Minor adjustments

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: docs/wiki/ModulesDesign.md [hunk 1]
```diff
@@ -155,9 +88,44 @@ module server_databases 'databases/deploy.bicep' = [for (database, index) in dat
 
 Each module should come with a `.bicep` folder with a least the `nested_cuaId.bicep` file in it
 
+## Naming
+
+Use the following naming standard for module files and folders:
+
+- Module folders are in camelCase and their name reflects the main resource type of the Bicep module they are hosting (e.g. `storageAccounts`, `virtualMachines`).
+- Cross-referenced and providers resource modules are placed in the `.bicep` subfolder and named `nested_<crossReferencedResourceType>.bicep`
```
Comment (MrMCake) [inline docs/wiki/ModulesDesign.md:96]:
```suggestion
- Cross-referenced and provider resource modules are placed in the `.bicep` subfolder and named `nested_<crossReferencedResourceType>.bicep`
```

File: docs/wiki/ModulesDesign.md [hunk 2]
```diff
@@ -414,17 +383,17 @@ Within a bicep file, follow the following conventions:
 
 ## Variables
 
-- camelCase, i.e `builtInRoleNames`
-- For modules that manage roleAssignments, update the list of roles to only be the applicable roles. You can find a helper script `Get-FormattedRBACRoles.ps1` in the `tools` folder of the repository.
+- Variable names are in camelCase, e.g. `builtInRoleNames`.
+- For modules that manage roleAssignments, typically cross-referenced `nested_rbac.bicep` in the `.bicep` folder, update the list of roles to only be the applicable roles. You can find a helper script `Get-FormattedRBACRoles.ps1` in the `utilities\tools\helper` folder of the repository.
```
Comment (MrMCake) [inline docs/wiki/ModulesDesign.md:387]:
```suggestion
- Modules that manage roleAssignments typically have a cross-referenced `nested_rbac.bicep` file in their `.bicep` folder.  Those files require you to update their list of roles to only the applicable roles. You can find a helper script `Get-FormattedRBACRoles.ps1` in the `utilities/tools/helper` folder of the repository.
```

File: docs/wiki/ModulesDesign.md [hunk 3]
```diff
@@ -474,34 +443,68 @@ Within a bicep file, follow the following conventions:
...
       }
       ```
-- Bicep `modules`:
-  - camel_Snake_Case, i.e `resourceGroup_rbac` ?
-  - File name for nested module is structured as follows: `nested_<resourceName>.bicep` i.e:
-    - `nested_rbac.bicep`
+## Modules
+
+  - Module symbolic names are in camel_Snake_Case, following the schema `<mainResourceType>_<referencedResourceType>` e.g. `storageAccount_fileServices`, `virtualMachine_nic`, `resourceGroup_rbac`.
+  - Modules enable you to reuse code from a Bicep file in other Bicep files. As such they're normally leveraged for deploying child resources (e.g. file services in a storage account), cross referenced resources (e.g. network interface in a virtual machine) or providers (e.g. role assignment in a resource group).
```
Comment (MrMCake) [inline docs/wiki/ModulesDesign.md:449]:
```suggestion
  - Modules enable you to reuse code from one Bicep file in another Bicep file. As such they're usually leveraged to deploy child resources (e.g. file services in a storage account), cross referenced resources (e.g. network interface in a virtual machine) or providers (e.g. role assignment in a resource group).
```

File: docs/wiki/ModulesDesign.md [hunk 4]
```diff
@@ -474,34 +443,68 @@ Within a bicep file, follow the following conventions:
...
+
+### Deployment names
+
+When using modules from parent resources you will need to specify a name that, when deployed, will be used to assign the deployment name.
```
Comment (MrMCake) [inline docs/wiki/ModulesDesign.md:453]:
```suggestion
When using modules from parent resources you will need to specify a name (i.e. deployment name) that, when deployed, will be used to assign the deployment name.
```

File: docs/wiki/ModulesDesign.md [hunk 5]
```diff
@@ -474,34 +443,68 @@ Within a bicep file, follow the following conventions:
...
+
+There are some constraints that needs to be considered when naming the deployment:
+
+- Deployment name length can't exceed 64 chars.
+- Two deployments with the same name created in different location will fail.
```
Comment (MrMCake) [inline docs/wiki/ModulesDesign.md:458]:
```suggestion
- Two deployments with the same name created in different Azure locations (e.g. WestEurope & EastUS) will fail.
```

File: docs/wiki/ModulesDesign.md [hunk 6]
```diff
@@ -474,34 +443,68 @@ Within a bicep file, follow the following conventions:
...
+- Using the same deployment name more than once, will surface only the last one in the Azure Portal.
```
Comment (MrMCake) [inline docs/wiki/ModulesDesign.md:459]:
```suggestion
- Using the same deployment name more than once, will surface only the most recent deployed one in the Azure Portal.
```

File: docs/wiki/ModulesDesign.md (hunk 1 above)
Comment (MariusStorhaug) [reply docs/wiki/ModulesDesign.md:96]: Extension resources... there is nothing in the official doc called provider resource modules.

File: docs/wiki/ModulesDesign.md (hunk 2 above)
Comment (MariusStorhaug) [reply docs/wiki/ModulesDesign.md:387]: I suggest moving this to the RBAC chapter on this article. Its not general consideration to variables, but this is more specific to RBAC.

File: docs/wiki/ModulesDesign.md (hunk 3 above)
Comment (MariusStorhaug) [reply docs/wiki/ModulesDesign.md:449]: Again there is nothing called providers. Call this extension resources

File: docs/wiki/ModulesDesign.md (hunk 1 above)
Comment (eriqua) [reply docs/wiki/ModulesDesign.md]: Replaced providers occurrences in this .md file. Probably we're using this naming in other files and they should be replaced as well.

File: docs/wiki/ModulesDesign.md (hunk 3 above)
Comment (eriqua) [reply docs/wiki/ModulesDesign.md]: Done.

File: docs/wiki/ModulesDesign.md (hunk 2 above)
Comment (eriqua) [reply docs/wiki/ModulesDesign.md]: Moved and updated

**Review by MrMCake — APPROVED** [review]

**Pull request merged**